src/hal/motor.cpp

```diff
@@ -2,20 +2,461 @@
+// Motor Position Control
+//*****************************************************************************
+
+// Set the current position as the zero point.
+void motor_position_zero()
+{
+  assert((motor_state != MOTOR_STATE_OPEN) && (motor_state != MOTOR_STATE_CLOSE));
+
+  uint8_t sreg = SREG; // save interrupt state
+  cli(); // disable interrupts
+
+  motor_control.microstep_position = 0;
+  motor_control.step_position = 0;
+  motor_control.step_command = 0;
+
+  SREG = sreg; // restore interrupt state
+}
+
+// Moves the motor to an angle relative to its current position.
+//
+// Parameters
+// ----------
+// int32_t angle      The angle (in degrees) to move the motor.
+// int8_t  direction  The direction to move the motor. One of:
+//                      * MOTOR_DIRECTION_OPEN
+//                      * MOTOR_DIRECTION_CLOSE
+void motor_move(int32_t angle, int8_t direction)
+{
+  // TODO: Do we want to have assertions? How are they implemented in this
+  // system? (i.e. Do they cause a watchdog reset, print a debug message to the
+  // console, etc.)
+  assert((direction == MOTOR_DIRECTION_OPEN) ||
+         (direction == MOTOR_DIRECTION_CLOSE));
+
+  uint8_t sreg = SREG; // save interrupt state
+  cli(); // disable interrupts
+  
+  motor_busy = true;
+  
+  // Convert to millidegrees and account for direction.
+
+  // Deriving motor_control.step_command from motor_control.step_position can
+  // cause error to accumulate from movement to movement. If
+  // motor_step_position is to be used for position control, then it needs  to
+  // be updated atomically in the motor position tracking ISR.
+  // motor_control.step_command = motor_control.step_position + direction*((angle*1000)/MOTOR_STEP_ANGLE);
+
+  motor_control.step_command += direction*((angle*1000)/MOTOR_STEP_ANGLE);
+
+  SREG = sreg; // restore interrupt state
+}
+
+// Moves the motor to the home position (fully open).
+void motor_home()
+{
+  // Move the motor far enough in the OPEN direction (360_, more than the total
+  // possible excursion) in order to guarantee that the top limit switch will
+  // be tripped.
+  motor_move(360, MOTOR_DIRECTION_OPEN);
+}
+
+// Moves the motor to the away position (fully closed).
+void motor_away() {
+  // Move the motor far enough in the CLOSE direction (360_, more than the
+  // total possible excursion) in order to guarantee that the bottom limit
+  // switch will be tripped.
+  motor_move(360, MOTOR_DIRECTION_CLOSE);
+}
+
+//*****************************************************************************
+// Interrupt Service Routines
+//*****************************************************************************
+
+// Motor State Control ISR
+static inline void motor_state_control_ISR()
+{
+  // Report position data every 0.5s.
+  DEBUG_PRINT_EVERY(50, "step_command:  %d", motor_control.step_command);
+  DEBUG_PRINT_EVERY(50, "step_position: %d", motor_control.step_position);
```

Comment:
See above on risk of hanging

Reply:
Is this really at risk of hanging if interrupts are disabled?

Reply:
Interrupts are disabled since youre in an ISR (unless you re-enable them to allow nesting interrupts via sei)

Reply:
TIL that AVR does not have nested interrupts by default.

Reply:
Removed debug prints.